### PR TITLE
[HGCAL] Tracksters to SimClusters associator

### PR DESCRIPTION
**Pull request by lecriste**
#### PR description:

This PR introduces an associator between SimClusters and TICL Tracksters. Among other usages, it will be used for HGCAL validation based on Tracksters in place of the current MultiClusters. No changes are expected in the output.

#### PR validation:

`runTheMatrix -l limited`

### CODE REVIEW

**Comment by civanch** [comment]
please test

File: DataFormats/HGCalReco/interface/Trackster.h [hunk 1]
```diff
@@ -199,5 +199,7 @@ namespace ticl {
     // trackster ID probabilities
     std::array<float, 8> id_probabilities_;
   };
+
+  typedef std::vector<Trackster> TracksterCollection;
```
Comment (slava77) [inline DataFormats/HGCalReco/interface/Trackster.h:203]: this was the only line for reco in #32941 
is this PR replacing #32941  or is there more?
From the reco POV, it looks the same now.

Comment (lecriste) [reply DataFormats/HGCalReco/interface/Trackster.h:203]: No, this PR is independent of #32941.
The PR containing the new data member for #32941 has been already merge (#33290).

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33333	bcac4c6
- code changes in reco category are essentially trivial, a `typedef std::vector<Trackster> TracksterCollection;` in DataFormats/HGCalReco/interface/Trackster.h

File: DataFormats/HGCalReco/interface/Trackster.h (hunk 1 above)
Comment (slava77) [reply DataFormats/HGCalReco/interface/Trackster.h:203]: sorry, the update is indeed different; even though in both cases it's a matter of one line for the reco category.

**Comment by lecriste** [comment]
@civanch can you please review this PR?

File: SimCalorimetry/HGCalAssociatorProducers/plugins/TracksterAssociatorByEnergyScoreImpl.cc [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+    }
+
+    float invTracksterEnergyWeight = 0.f;
+    for (const auto& lcId : tracksters[tsId].vertices()) {
+      const auto& hits_and_fractions = layerClusters[lcId].hitsAndFractions();
+
+      // Compute the correct normalization
+      for (auto const& haf : hits_and_fractions) {
+        invTracksterEnergyWeight +=
+            (haf.second * hitMap_->at(haf.first)->energy()) * (haf.second * hitMap_->at(haf.first)->energy());
```
Comment (rovere) [inline SimCalorimetry/HGCalAssociatorProducers/plugins/TracksterAssociatorByEnergyScoreImpl.cc:339]: Aren't we missing the *multiplicity* usage of the `layerCluster` among different `tracksters`?
Here and, possibly, in other places in this `associator`?

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33333	886d649
- there were no changes in reco code since the last signoff

**Comment by civanch** [comment]
+1

File: SimDataFormats/Associations/plugins/BuildFile.xml [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <use name="FWCore/Framework"/>
 <use name="FWCore/ParameterSet"/>
+<use name="eigen"/>
```
Comment (davidlange6) [inline SimDataFormats/Associations/plugins/BuildFile.xml:3]: hi @lecriste - what triggers the need for this dependency?

Comment (lecriste) [reply SimDataFormats/Associations/plugins/BuildFile.xml:3]: Hi @davidlange6, that dependency is needed by [this include](https://github.com/cms-sw/cmssw/blob/886d64926c230c57bea1d6c1fdc72e72c19bf497/DataFormats/HGCalReco/interface/Trackster.h#L12).

Comment (davidlange6) [reply SimDataFormats/Associations/plugins/BuildFile.xml:3]: So then its the wrong change. Instead, add a dependency on DataFormats/HGCalReco. Looks like you have also added DataFormats/ForwardDetId and RecoLocalCalo/HGCalRecAlgos that should be added to the BuildFIle.

Comment (lecriste) [reply SimDataFormats/Associations/plugins/BuildFile.xml]: OK, fixed in the latest commit.

**Comment by davidlange6** [comment]
please test

File: SimDataFormats/Associations/plugins/BuildFile.xml [hunk 4]
```diff
@@ -1,5 +1,8 @@
 <use name="FWCore/Framework"/>
 <use name="FWCore/ParameterSet"/>
+<use name="DataFormats/HGCalReco"/>
+<use name="DataFormats/ForwardDetId"/>
+<use name="RecoLocalCalo/HGCalRecAlgos"/>
```
Comment (slava77) [inline SimDataFormats/Associations/plugins/BuildFile.xml:5]: does https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Restrictions_on_the_package_defi
still apply to  SimDataFormats/Associations?

One case/example is a plugins/ is not expected

Comment (davidlange6) [reply SimDataFormats/Associations/plugins/BuildFile.xml:5]: its true that SimDataFormats/Associations having a plugins directory doesn't make much sense. That is already the case, however

Comment (slava77) [reply SimDataFormats/Associations/plugins/BuildFile.xml:5]: but "already the case" seems to be a recent addition #32427 (not something from long time ago)

**Comment by civanch** [comment]
+1

File: SimDataFormats/Associations/plugins/BuildFile.xml (hunk 4 above)
Comment (slava77) [reply SimDataFormats/Associations/plugins/BuildFile.xml:5]: can this plugin be decoupled from the DataFormats now or is there a strong preference to do it in a follow up PR?

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33333 087cef2 
- since the last reco signoff the updates were outside of the reco category. 
- presence of a plugins directory in SimDataFormats/Associations seems questionable to me, but apparently sim signed off on this

**Comment by lecriste** [comment]
@cms-sw/upgrade-l2 can you please review this PR?

**Comment by srimanob** [comment]
+Upgrade

**Comment by silviodonato** [comment]
+1

@lecriste please have a look at the comments https://github.com/cms-sw/cmssw/pull/33333#discussion_r610591196

**Pull request merged**